src/rp2_common/hardware_pio/pio.c

```diff
@@ -42,8 +42,9 @@ void pio_sm_unclaim(PIO pio, uint sm) {
 }
 
 int pio_claim_unused_sm(PIO pio, bool required) {
-    uint which = pio_get_index(pio);
-    uint base = which * NUM_PIO_STATE_MACHINES;
+    // PIO index is 0 or 1.
+    int which = (int)pio_get_index(pio);
+    int base = which * NUM_PIO_STATE_MACHINES;
     int index = hw_claim_unused_from_range((uint8_t*)&claimed, required, base,
                                       base + NUM_PIO_STATE_MACHINES - 1, "No PIO state machines are available");
     return index >= base ? index - base : -1;
```

Comment:
Experimentally, it seems you can leave `which` and `base` as `uint`, and then just change this last line to:
`return index >= (int)base ? index - (int)base : -1;`

(but perhaps that's actually not better than the version you've proposed :man_shrugging: )